apollo-router-core/src/cache.rs

```diff
@@ -160,15 +169,43 @@ mod tests {
         }
     }
 
+    mock! {
+        HasACacheResolver {}
+
+        #[async_trait]
+        impl CacheResolver<usize, usize> for HasACacheResolver {
+            async fn retrieve(&self, key: usize) -> Result<usize, CacheResolverError>;
+        }
+    }
+
     #[test(tokio::test)]
     async fn it_should_enforce_cache_limits() {
-        let cache = HasACache::new(13);
+        let cache = HasACache::new(Box::new(HasACacheResolver {}), 13);
 
         for i in 0..14 {
-            cache.cm.get(i).await.expect("gets the value");
+            cache.get(i).await.expect("gets the value");
         }
         let guard = cache.cm.cached.lock().await;
-        println!("{:?}", guard);
         assert_eq!(guard.len(), 13);
     }
+
+    #[test(tokio::test)]
+    async fn it_should_only_delegate_once_per_key() {
+        let mut mock = MockHasACacheResolver::new();
+
+        mock.expect_retrieve().times(1).return_const(Ok(1));
+
+        let cache = HasACache::new(Box::new(mock), 10);
+
+        // Let's trigger 100 concurrent gets of the same value and ensure only
+        // one delegated retrieve is made
+        let computations = (0..100).map(|_| cache.get(1));
+        let _ = join_all(computations)
```

Comment:
join_all is not a good way to test concurrency, it tries its list of futures in order on every `poll()` call: https://docs.rs/futures-util/0.3.18/src/futures_util/future/join_all.rs.html#142

Reply:
I thought it drove all futures concurrently. The documentation certainly implies that: https://docs.rs/futures/0.3.18/futures/future/fn.join_all.html

Reply:
it depends on the order in which they wake up. But since all of them execute the same code from inside the same thread it will result in a very predictable execution

Reply:
is there a way to add a delay within the mock's return function ? Would that work ?

Reply:
@Geal I can replace that code with:```
        let mut computations: FuturesUnordered<_> = (0..100).map(|_| cache.get(1)).collect();

        while let Some(result) = computations.next().await {
            result.expect("result retrieved");
        }
```
That should be as concurrent as we can possibly have it...

@o0Ignition0o I don't think we want a delay, but I can look into adding that if we think it improves the tests.

Reply:
I have no idea to be honest, it's just a random suggestion ^^

Reply:
I'm considering this resolved by the use of FuturesUnordered.
Done in commit: [6a63ab412a7b4fb6c8b9ab76520dfb6d4ca785c0](https://github.com/apollographql/router/pull/235/commits/6a63ab412a7b4fb6c8b9ab76520dfb6d4ca785c0)